### PR TITLE
Support speculative decoding in the trtllm_mha attention backend

### PR DESCRIPTION
**Pull request by Qiaolin-Yu**
### Launch the server
```
python3 -m sglang.launch_server   --model openai/gpt-oss-120b   --speculative-algo EAGLE3   --speculative-draft zhuyksir/EAGLE3-gpt-oss-120b-bf16   --speculative-num-steps 2   --speculative-eagle-topk 1   --speculative-num-draft-tokens 3    --dtype float16 --attention-backend trtllm_mha    --tp 4 --port 40010
```

### Run the accuracy test
```
OPENAI_API_KEY=dummy python -m gpt_oss.evals --base-url http://localhost:40010/v1 --model dummy --reasoning-effort medium --eval gpqa --n-threads 1000
```

### Results
{'chars': np.float64(69.8705808080808), 'chars:std': np.float64(227.15709320445401), 'score': np.float64(0.7146464646464646), 'score:std': np.float64(0.4515826560163422)}

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).

### CODE REVIEW

File: python/sglang/srt/layers/attention/trtllm_mha_backend.py [hunk 1]
```diff
@@ -99,17 +115,81 @@ def init_cuda_graph_state(
         """Initialize CUDA graph state for TRTLLM MHA."""
         self.decode_cuda_graph_metadata = {
             "cache_seqlens": torch.zeros(max_bs, dtype=torch.int32, device=self.device),
+            "cu_seqlens_q": torch.arange(
```
Comment (Fridge003) [inline python/sglang/srt/layers/attention/trtllm_mha_backend.py]: I guess `cu_seqlens_q`, `cu_seqlens_k` and `page_table_draft_decode` are not used when speculative decoding is disabled? So we might only create them when spec is enabled.

**Comment by qywu** [comment]
I am getting lower than expected accpet rate when using trtllm_mha attention backend for Eagle3 with cuda graph.

Here is my launch command:
```bash
python -m sglang.launch_server \
  --model-path meta-llama/Llama-3.1-8B-Instruct \
  --port 8005 \
  --tensor-parallel-size 1 \
  --speculative-algo EAGLE3 \
  --speculative-draft jamesliu1/sglang-EAGLE3-Llama-3.1-Instruct-8B \
  --speculative-num-steps 3 \
  --speculative-eagle-topk 1 \
  --speculative-num-draft-tokens 4 \
  --mem-fraction 0.6 \
  --attention-backend trtllm_mha \
  --dtype bfloat16
```

image

For comparison, when I disable cuda graph,
```bash
python -m sglang.launch_server \
  --model-path meta-llama/Llama-3.1-8B-Instruct \
  --port 8005 \
  --tensor-parallel-size 1 \
  --speculative-algo EAGLE3 \
  --speculative-draft jamesliu1/sglang-EAGLE3-Llama-3.1-Instruct-8B \
  --speculative-num-steps 3 \
  --speculative-eagle-topk 1 \
  --speculative-num-draft-tokens 4 \
  --mem-fraction 0.6 \
  --attention-backend trtllm_mha \
  --dtype bfloat16 \
  --disable-cuda-graph
```

image


Also, with flashinfer,
```bash
python -m sglang.launch_server \
  --model-path meta-llama/Llama-3.1-8B-Instruct \
  --port 8005 \
  --tensor-parallel-size 1 \
  --speculative-algo EAGLE3 \
  --speculative-draft jamesliu1/sglang-EAGLE3-Llama-3.1-Instruct-8B \
  --speculative-num-steps 3 \
  --speculative-eagle-topk 1 \
  --speculative-num-draft-tokens 4 \
  --mem-fraction 0.6 \
  --attention-backend flashinfer \
  --dtype bfloat16
```

image

---

This is the prompt I used for testing:
```bash
curl http://localhost:8005/v1/chat/completions   -H "Content-Type: application/json"   -d '{
    "model": "",
    "messages": [
      {
        "role": "user",
        "content": "What is flash attention"
      }
    ],
    "max_tokens": 1024,
    "temperature": 0.0
  }'
```

**Comment by Qiaolin-Yu** [comment]
> I am getting lower than expected accpet rate when using trtllm_mha attention backend for Eagle3 with cuda graph.
> 
> Here is my launch command:
> 
> ```shell
> python -m sglang.launch_server \
>   --model-path meta-llama/Llama-3.1-8B-Instruct \
>   --port 8005 \
>   --tensor-parallel-size 1 \
>   --speculative-algo EAGLE3 \
>   --speculative-draft jamesliu1/sglang-EAGLE3-Llama-3.1-Instruct-8B \
>   --speculative-num-steps 3 \
>   --speculative-eagle-topk 1 \
>   --speculative-num-draft-tokens 4 \
>   --mem-fraction 0.6 \
>   --attention-backend trtllm_mha \
>   --dtype bfloat16
> ```
> 
> image
> For comparison, when I disable cuda graph,
> 
> ```shell
> python -m sglang.launch_server \
>   --model-path meta-llama/Llama-3.1-8B-Instruct \
>   --port 8005 \
>   --tensor-parallel-size 1 \
>   --speculative-algo EAGLE3 \
>   --speculative-draft jamesliu1/sglang-EAGLE3-Llama-3.1-Instruct-8B \
>   --speculative-num-steps 3 \
>   --speculative-eagle-topk 1 \
>   --speculative-num-draft-tokens 4 \
>   --mem-fraction 0.6 \
>   --attention-backend trtllm_mha \
>   --dtype bfloat16 \
>   --disable-cuda-graph
> ```
> 
> image
> Also, with flashinfer,
> 
> ```shell
> python -m sglang.launch_server \
>   --model-path meta-llama/Llama-3.1-8B-Instruct \
>   --port 8005 \
>   --tensor-parallel-size 1 \
>   --speculative-algo EAGLE3 \
>   --speculative-draft jamesliu1/sglang-EAGLE3-Llama-3.1-Instruct-8B \
>   --speculative-num-steps 3 \
>   --speculative-eagle-topk 1 \
>   --speculative-num-draft-tokens 4 \
>   --mem-fraction 0.6 \
>   --attention-backend flashinfer \
>   --dtype bfloat16
> ```
> 
> image
> This is the prompt I used for testing:
> 
> ```shell
> curl http://localhost:8005/v1/chat/completions   -H "Content-Type: application/json"   -d '{
>     "model": "",
>     "messages": [
>       {
>         "role": "user",
>         "content": "What is flash attention"
>       }
>     ],
>     "max_tokens": 1024,
>     "temperature": 0.0
>   }'
> ```

Thanks! I will check this.

File: python/sglang/srt/layers/attention/trtllm_mha_backend.py (hunk 1 above)
Comment (hebiao064) [reply python/sglang/srt/layers/attention/trtllm_mha_backend.py]: good point: cu_seqlens_q, cu_seqlens_k should be needed
page_table_draft_decode can be set when spec is enabled

Comment (hebiao064) [reply python/sglang/srt/layers/attention/trtllm_mha_backend.py]: https://github.com/sgl-project/sglang/pull/9455

File: python/sglang/srt/layers/attention/trtllm_mha_backend.py [hunk 2]
```diff
@@ -99,17 +117,81 @@ def init_cuda_graph_state(
...
+            ),
             "page_table": torch.zeros(
                 max_bs,
                 (self.max_context_len + self.page_size - 1) // self.page_size,
                 dtype=torch.int32,
                 device=self.device,
             ),
+            "page_table_draft_decode": torch.zeros(
+                max_bs,
+                (self.max_context_len + self.page_size - 1) // self.page_size,
```
Comment (hebiao064) [inline python/sglang/srt/layers/attention/trtllm_mha_backend.py]: try `max_num_pages = (self.max_context_len + self.page_size - 1) // self.page_size` and reuse `max_num_pages`

Comment (Qiaolin-Yu) [reply python/sglang/srt/layers/attention/trtllm_mha_backend.py]: done

File: python/sglang/srt/layers/attention/trtllm_mha_backend.py (hunk 1 above)
Comment (Qiaolin-Yu) [reply python/sglang/srt/layers/attention/trtllm_mha_backend.py]: done

**Comment by zhyncs** [comment]
follow up @Qiaolin-Yu
